lib/core/src/display/rendering/htmlrenderer.rs

```diff
@@ -22,32 +34,48 @@ impl HTMLRenderer {
 
     /// Renders the `Scene` from `Camera`'s point of view.
     pub fn render(&self, camera: &mut Camera, scene: &HTMLScene) {
-        // Note [znear from projection matrix]
-        let half_dim  = scene.get_dimensions() / 2.0;
-        let expr      = camera.projection[(1, 1)];
-        let near      = format!("{}px", expr * half_dim.y);
         let trans_cam = camera.transform.to_homogeneous().try_inverse();
         let trans_cam = trans_cam.expect("Camera's matrix is not invertible.");
         let trans_cam = trans_cam.map(eps);
         let trans_cam = invert_y(trans_cam);
-        let trans_z   = format!("translateZ({})", near);
-        let matrix3d  = trans_cam.into_css_matrix();
-        let trans     = format!("translate({}px,{}px)", half_dim.x, half_dim.y);
-        let css       = format!("{} {} {}", trans_z, matrix3d, trans);
 
-        scene.div   .element.set_property_or_panic("perspective", near);
-        scene.camera.element.set_property_or_panic("transform"  , css);
+        // Note [znear from projection matrix]
+        let half_dim     = scene.get_dimensions() / 2.0;
+        let y_scale      = camera.get_y_scale();
+        let near         = (y_scale * half_dim.y).into();
+        let half_width   = half_dim.x.into();
+        let half_height  = half_dim.y.into();
+        let scene_dom    = &scene.html_data.div.element;
+        let camera_dom   = &scene.html_data.camera.element;
+
+        // Note [Unsafe Float32ArrayView]
+        let matrix_array = unsafe { trans_cam.into_float32_array_view() };
 
-        for object in &scene.objects {
+        setup_perspective(&scene_dom, &near);
+        setup_camera_transform(
+            &camera_dom,
+            &near,
+            &half_width,
+            &half_height,
+            &matrix_array
+        );
+
+        for object in scene {
             let mut transform = object.transform.to_homogeneous();
             transform.iter_mut().for_each(|a| *a = eps(*a));
-            let matrix3d = transform.into_css_matrix();
-            let css      = format!("translate(-50%, -50%) {}", matrix3d);
-            object.element.set_property_or_panic("transform", css);
+
+            // Note [Unsafe Float32ArrayView]
+            let matrix_array = unsafe { transform.into_float32_array_view() };
+            set_object_transform(&object.element, &matrix_array);
```

Comment:
This code is error prone. It's too easy to insert a line between these lines and make it break. Please put inside `unsafe` block the whole unsafe code and refactor it to a safe function. Document the function extensively. Documentation of each function constining unsafe block should be documented using the special Safety / Unsafety sections. We should not use `unsafe` call as a part of safe code. The idea is to refactor unsafe fragments to self-contained safe functions which are guaranteed to work correctly.

Reply:
I created a safe function wrapping the unsafe code :+1: